addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,219 @@
+
+        for (int i = 0; i < 4; i++) {
+            _update(quad->get_child(i), lod - 1, view_pos);
+            if (quad->get_child(i)->has_children())
+                no_split_child = false;
+        }
+
+        if (no_split_child && world_center.distance_to(view_pos) > split_distance) {
+            // Join
+            if (quad->has_children()) {
+                for (int i = 0; i < 4; i++) {
+                    Quad* child = quad->get_child(i);
+                    _recycle_chunk(child->data, child->origin_x, child->origin_y, lod - 1);
+                    quad->data = Variant();
+                }
+                quad->clear_children();
+
+                //assert(quad->data == null);
+                //if ( quad->data.is_valid() )    // Really necessary after it was just unrefed in quad?
+                	//Godot::print("QuadTreeLod error: quad->data is still valid");
+
+                quad->data = _make_chunk(lod, quad->origin_x, quad->origin_y);
+            }
+        }
+    }
+} // _update
+
+void QuadTreeLod::_join_all_recursively(Quad *quad, int lod) {
+    if (quad->has_children()) {
+        for (int i = 0; i < 4; i++) {
+            _join_all_recursively(quad->get_child(i), lod - 1);
+        }
+        quad->clear_children();
+
+    } else if (quad->data.get_type() == Variant::OBJECT) {
+        _recycle_chunk(quad->data, quad->origin_x, quad->origin_y, lod);
+        quad->data = Variant();
+    }
+}
+
+Variant QuadTreeLod::_make_chunk(int lod, int origin_x, int origin_y) {
+    if (_make_func.is_valid())
+        return _make_func->call_func(origin_x, origin_y, lod);
+    else
+        return Variant();
+}
+
+void QuadTreeLod::_recycle_chunk(Variant chunk, int origin_x, int origin_y, int lod) {
+    if (_recycle_func.is_valid())
+        _recycle_func->call_func(chunk, origin_x, origin_y, lod);
+}
+
+void QuadTreeLod::debug_draw_tree(Variant ci) {
+    if (ci.get_type() == Variant::OBJECT && ci.has_method("draw_rect"))
+        _debug_draw_tree_recursive(ci, _tree, _max_depth, 0);
+}
+
+void QuadTreeLod::_debug_draw_tree_recursive(Variant ci, Quad *quad, int lod_index, int child_index) {
+    if (quad->has_children()) {
+        for (int i = 0; i < 4; i++) {
+            _debug_draw_tree_recursive(ci, quad->get_child(i), lod_index - 1, i);
+        }
+
+    } else {
+        int size = get_lod_factor(lod_index);
+        int checker = 0;
+        if (child_index == 1 || child_index == 2)
+            checker = 1;
+
+        int chunk_indicator = 0;
+        if (quad->data.get_type() == Variant::OBJECT)
+            chunk_indicator = 1;
+
+        Variant rect2(Rect2( Vector2((real_t)quad->origin_x, (real_t)quad->origin_y) * (real_t)size, 
+                             Vector2((real_t)size, (real_t)size)));
+        Variant color(Color(1.0 - (real_t)lod_index * 0.2, 0.2 * (real_t)checker, (real_t)chunk_indicator, 1));
+        Variant *args[] = { &rect2, &color };
+        ci.call("draw_rect", (const Variant **)args, 2);
```

Comment:
Instead of using `call` which allocates a string everytime and looks up the method dynamically, you may expect a `CanvasItem*` instead of `Variant` so you can call it directly